### PR TITLE
Optional Init Definition

### PR DESCRIPTION
**Pull request by ppaulweber**
- moves the strict `program` function checking from the `ConsistencyCheckPass` to the `FrameSizeDeterminationPass`
- makes access and evaluation to the internal `program` function in several other passes optional
- related to casm-lang/casm#40

### CODE REVIEW

File: src/analyze/FrameSizeDeterminationPass.cpp [hunk 1]
```diff
@@ -108,19 +110,40 @@ class FrameSizeDeterminationVisitor final : public RecursiveVisitor
...
+, m_foundProgramFunction( false )
 {
 }
 
+void FrameSizeDeterminationVisitor::visit( Specification& node )
+{
+    node.header()->accept( *this );
+    node.definitions()->accept( *this );
+
+    if( not m_foundProgramFunction )
```
Comment (emmanuel099) [inline src/analyze/FrameSizeDeterminationPass.cpp]: These changes don't belong here (i.e. have nothing in common with "frame size calculation").

File: src/analyze/SymbolResolverPass.cpp [hunk 2]
```diff
@@ -128,7 +128,11 @@ SymbolResolveVisitor::SymbolResolveVisitor( libcasm_fe::Logger& log, Namespace&
 void SymbolResolveVisitor::visit( InitDefinition& node )
 {
     RecursiveVisitor::visit( node );
-    node.programFunction()->accept( *this );
+    const auto programFunction = node.programFunction();
+    if( programFunction )
```
Comment (emmanuel099) [inline src/analyze/SymbolResolverPass.cpp]: Always true, because InitDefition => programFunction

File: src/analyze/TypeCheckPass.cpp [hunk 3]
```diff
@@ -96,7 +96,11 @@ TypeCheckVisitor::TypeCheckVisitor( libcasm_fe::Logger& log, Namespace& symbolta
 void TypeCheckVisitor::visit( InitDefinition& node )
 {
     RecursiveVisitor::visit( node );
-    node.programFunction()->accept( *this );
+    const auto programFunction = node.programFunction();
+    if( programFunction )
```
Comment (emmanuel099) [inline src/analyze/TypeCheckPass.cpp]: Always true, because InitDefition => programFunction

File: src/analyze/TypeInferencePass.cpp [hunk 4]
```diff
@@ -173,7 +173,11 @@ TypeInferenceVisitor::TypeInferenceVisitor( libcasm_fe::Logger& log, const Names
 void TypeInferenceVisitor::visit( InitDefinition& node )
 {
     RecursiveVisitor::visit( node );
-    node.programFunction()->accept( *this );
+    const auto programFunction = node.programFunction();
+    if( programFunction )
```
Comment (emmanuel099) [inline src/analyze/TypeInferencePass.cpp]: Always true, because InitDefition => programFunction

File: src/analyze/FrameSizeDeterminationPass.cpp (hunk 1 above)
Comment (ppaulweber) [reply src/analyze/FrameSizeDeterminationPass.cpp]: yes you are right, I've removed this check and fixed this directly in the `NumericExecutionPass`

**Comment by ppaulweber** [comment]
@emmanuel099 please review again!

File: src/execute/NumericExecutionPass.cpp [hunk 5]
```diff
@@ -2078,6 +2113,13 @@ u1 NumericExecutionPass::run( libpass::PassResult& pr )
         InvariantChecker invariantChecker( locationRegistry, globalState );
         invariantChecker.check( *specification );
 
+        if( not scheduler.check() )
```
Comment (emmanuel099) [inline src/execute/NumericExecutionPass.cpp]: Move this warning after the while loop, and use `scheduler.numberOfSteps() == 0` condition instead. The `check` method can then be dropped.

Also another warning message would be good, because the current one is not 100% correct, e.g. `init {(0)-> undef}`. -> there is an init definition, but no agent is initially available.

Maybe: "Could not perform a single step because no agent was initially available. This may happen when no valid initial rule has been specified (see InitRule)."

**Comment by ppaulweber** [comment]
> Use `rebase` to drop/cleanup changes instead of manually reverting them (make log, blame, ... unusable).
> 
> https://github.com/mockito/mockito/wiki/Using-git-to-prepare-your-PR-to-have-a-clean-history

you are right again, I've to add this to the `CONTRIBUTION` guidelines, but for this PR it will stay as is

File: src/execute/NumericExecutionPass.cpp (hunk 5 above)
Comment (ppaulweber) [reply src/execute/NumericExecutionPass.cpp]: good point, could you please change this so that we can finalize this PR? thx

Comment (emmanuel099) [reply src/execute/NumericExecutionPass.cpp]: The `scheduler.numberOfSteps() == 0` is correct in this cause, because when no agents are available the scheduler doesn't make a step, see https://github.com/casm-lang/libcasm-fe/blob/master/src/execute/NumericExecutionPass.cpp#L1888

Comment (ppaulweber) [reply src/execute/NumericExecutionPass.cpp]: true, but the problem and therefore the implementation of the `check` was that the `collectAgents` method results into an error for not picking up a `IR::ReferenceConstant` because the "no init" results into a `VoidConstant`, just try it out, you will see :smile:

**Comment by ppaulweber** [comment]
@emmanuel099 please provide the last fix and assign me to review it

**Comment by ppaulweber** [comment]
@emmanuel099 according to the CI there is a segmentation fault in the Windows and Darwin builds, any ideas why?

File: etc/test/passes.cpp [hunk 6]
```diff
@@ -97,7 +97,16 @@ SOURCE_TEST( analyze, ConsistencyCheckPass, source, true, , );
 SOURCE_TEST( transform, AstDumpSourcePass, source, true, , );
 SOURCE_TEST( transform, AstDumpDotPass, source, true, , );
 
-// NumericExecutionPass
+SOURCE_TEST( execute, NumericExecutionPass, source, true, , );
```
Comment (emmanuel099) [inline etc/test/passes.cpp]: The problem happens in this test case.

Trace-output (just added some couts):
```
StateInitializationVisitor::Specification
StateInitializationVisitor::InitDefinition
StateInitializationVisitor::FunctionDefinition
ExecutionVisitor::ReferenceLiteral test_skip
ExecutionVisitor::UpdateRule [v] @test_skip
StateInitializationVisitor::Specification updates fired!
AgentScheduler::step
AgentScheduler::collectAgents
ExecutionVisitor::collectAgents::loop [Agent] @c0 = Agent Agent -> [v] v undef
libcasm-fe-check: casm-fe/src/execute/NumericExecutionPass.cpp:1940: std::vector<Agent> AgentScheduler::collectAgents() const: Assertion `IR::isa< ReferenceConstant >( value )' failed.
```
Somewhere `[v] @test_skip` becomes `[v] v undef`. But this also happens in master, so this test case should be moved to another PR to fix the problem there.

File: src/execute/NumericExecutionPass.cpp (hunk 5 above)
Comment (emmanuel099) [reply src/execute/NumericExecutionPass.cpp]: Which means that `check` has hidden the real problem. If you get a void constant where a reference is expected means that type preservation is violated.

File: etc/test/passes.cpp (hunk 6 above)
Comment (ppaulweber) [reply etc/test/passes.cpp]: @emmanuel099 :+1: in this case I will open a bug with this issue and revert this change in the `etc/test/passes.cpp` unit test case file

**Comment by ppaulweber** [comment]
@emmanuel099 for this PR the `codecov/project — 26.48% (-0.17%) compared to dd60691` can be ignored, please re-review and approve if everything is OK, thx!

File: src/execute/NumericExecutionPass.cpp (hunk 5 above)
Comment (emmanuel099) [reply src/execute/NumericExecutionPass.cpp]: There went something wrong during rebase. Dropped commits have been re-added and these changes (+ unrelated changes in FrameSizeDeterminationVisitor) are included again.

**Comment by ppaulweber** [comment]
@emmanuel099 that is true, but you already have approved and merged casm-lang/libcasm-tc#78 and this breaks now the master upstream. We will skip the mentioned rebase and additional fixes are done in an new issue/pr.
Therefore I'll merge this now.
